src/test/java/com/google/sps/servlets/SearchServletTest.java

```diff
@@ -0,0 +1,151 @@
+// // Copyright 2019 Google LLC
+// //
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.when;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.http.HttpServletRequest;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+/** */
+public final class SearchServletTest {
+
+  private static final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+  private SearchServlet searchObject;
+
+  @BeforeEach
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+    searchObject = new SearchServlet();
+  }
+
+  @AfterEach
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Mock HttpServletRequest request;
+
+  @Test
+  /* Ensure the whole doGet process works given all search filters are set. */
+  public void GetCourses_AllParamsSet() throws IOException {
+
+    request =
+        createRequest(
+            request, /*name*/
+            "CS 105", /*professor*/
+            "Smith", /*term*/
+            "Spring 2020", /*units*/
+            "1,2,3");
+    List<Course> expectedCourses = new ArrayList<>();
+
+    addCourseEntity("CS 105", "Smith", "Spring 2020", 1);
+    expectedCourses.add(new Course("CS 105", "Smith", Long.valueOf(1), "Spring 2020"));
+
+    addCourseEntity("CS 105", "Smith", "Spring 2020", 2);
+    expectedCourses.add(new Course("CS 105", "Smith", Long.valueOf(2), "Spring 2020"));
+
+    addCourseEntity("CS 106", "Smith", "Spring 2020", 3);
+
+    List<Course> courses = searchObject.getHelper(request);
```

Comment:
Ideally you only want to test your class's public API. In this case that would be `doGet` (and `doPost`). You only want to resort to making a helper function like this public for testing if there really isn't a way to directly test your public API.

Reply:
Hmmm dang okay, I'm running into a weird NullPointerException with my response when trying to test doGet directly - apparently Dagm and Laura also ran into the same issue and talked to Travis about it and are just kinda doing the same thing with the helper function that does the same thing as doGet

Reply:
Yeah the Java strict typing makes this a bit of a nightmare. Dagm and I tried a few things then I told him to just use the helper, because it wasn't worth wasting hours on. 
Good enough to know that typically you want to test the public API

Reply:
Ah, if that's the way things are then it can't be helped, carry on!